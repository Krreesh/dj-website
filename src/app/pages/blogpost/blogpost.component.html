<div class="container-fluid pt-4 pb-4">
  <h1>{{'blogpost.title' | translate}}</h1>
  <blogpost-overview></blogpost-overview>
</div>
