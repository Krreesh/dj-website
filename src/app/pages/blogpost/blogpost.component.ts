import {Component} from '@angular/core';

@Component({
  selector: 'app-data',
  template: `
    <div class="container-fluid pt-4 pb-4">
      <h1>{{'blogpost.title' | translate}}</h1>
      <blogpost-overview></blogpost-overview>
    </div>
  `
})
export class BlogpostComponent {

}
