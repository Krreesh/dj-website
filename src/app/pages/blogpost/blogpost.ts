export interface Blogpost {
  title: string;
  publicationDate: Date;
  url: string;
  category: string;
  base64image: string;
}
