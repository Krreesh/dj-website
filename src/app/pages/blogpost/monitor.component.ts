import {ChangeDetectionStrategy, Component, ElementRef, Input, ViewChild} from '@angular/core';

@Component({
  selector: 'app-monitor',
  template: `
    <div class="monitor" #monitor (scroll)="marginTop = monitor.scrollTop">
      <a href="{{ url }}">
        <img
          class="blogpost"
          defaultImage="assets/images/placeholder.png"
          [lazyLoad]="'data:image/jpg;base64,' + image"
          alt="Preview of {{ title }}"
        />
        <img
          class="blogpost-view"
          src="assets/images/eye.png"
          [ngStyle]="{ top: marginTop + 'px' }"
        />
      </a>
    </div>
    <div class="monitor-stand"></div>
    <div class="monitor-stand-foot"></div>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush,
  styles: [
      `
      .blogpost {
        position: absolute;
        height: 200px;
        width: 170px;
      }

      .blogpost-view {
        opacity: 0;
        position: absolute;
        height: 130px;
        width: 170px;
        z-index: 1;
      }

      .blogpost-view:hover {
        opacity: 60%;
      }

      .monitor {
        position: relative;
        width: 200px;
        height: 150px;
        overflow-y: scroll;
        border: solid 10px black;
        border-radius: 0.5em;
      }

      .monitor::-webkit-scrollbar {
        width: 10px;
      }

      .monitor::-webkit-scrollbar-thumb {
        background: dimgray;
      }

      ::-webkit-scrollbar-track {
        background-color: lightgrey;
      }

      .monitor-stand {
        background-color: darkgray;
        height: 20px;
        margin-left: 60px;
        margin-right: 60px;
      }

      .monitor-stand-foot {
        background-color: #373535;
        border-top-left-radius: 0.5em;
        border-top-right-radius: 0.5em;
        height: 10px;
        margin-left: 40px;
        margin-right: 40px;
      }
    `
  ]
})
export class MonitorComponent {
  @Input()
  url: string;

  @Input()
  image: string;

  @Input()
  title: string;

  @ViewChild('monitor')
  monitor: ElementRef;

  marginTop = 0;
}
