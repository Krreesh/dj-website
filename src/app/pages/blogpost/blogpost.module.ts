import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {BlogpostsComponent} from './blogposts.component';
import {TranslateModule} from '@ngx-translate/core';
import {NgbModule} from '@ng-bootstrap/ng-bootstrap';
import {DataModule} from '../data/data.module';
import {DateModule} from '../../common/date.module';
import {BlogpostComponent} from './blogpost/blogpost.component';

@NgModule({
  imports: [
    CommonModule,
    TranslateModule,
    NgbModule,
    DataModule,
    DateModule
  ],
  declarations: [
    BlogpostComponent,
    BlogpostsComponent
  ]
})
export class BlogpostModule {

}
