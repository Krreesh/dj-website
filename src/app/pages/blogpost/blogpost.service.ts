import {Injectable} from '@angular/core';
import {combineLatest, Observable} from 'rxjs';
import {Blogpost} from './blogpost';
import {HttpClient, HttpHeaders} from '@angular/common/http';
import {TranslateServiceFacade} from '../../translation/translate.service.facade';
import {filter, map, tap} from 'rxjs/operators';

@Injectable({
  providedIn: 'root'
})
export class BlogpostService {
  private untranslatedBlogposts: Blogpost[];
  private readonly translatedBlogposts$: Observable<Blogpost[]>;
  private readonly baseUrl = `http://www.dieterjordens.be:10002/api/blogposts`;

  constructor(
    private http: HttpClient,
    private translateService: TranslateServiceFacade
  ) {
    const currentLanguage$ = translateService.getCurrentLanguage$();
    const httpOptions = {
      headers: new HttpHeaders({
        'Content-Type': 'application/json',
        Authorization: 'Basic YWRtaW46RWVuRWVudm91ZGlnV2FjaHR3b29yZA=='
      })
    };
    const blogposts$: Observable<Blogpost[]> = this.http.get<Blogpost[]>(
      this.baseUrl,
      httpOptions
    );

    this.translatedBlogposts$ = combineLatest([
      blogposts$,
      currentLanguage$
    ]).pipe(
      filter(([, currentLanguage]) => !!currentLanguage),
      map(([blogposts]) => blogposts),
      tap((blogposts) => this.cacheBlogposts(blogposts)),
      map(() => this.translateBlogposts())
    );
  }

  getBlogposts$(): Observable<Blogpost[]> {
    return this.translatedBlogposts$;
  }

  private cacheBlogposts(blogposts: Blogpost[]) {
    if (!this.untranslatedBlogposts) {
      this.untranslatedBlogposts = blogposts;
    }
  }

  private translateBlogposts(): Blogpost[] {
    return this.untranslatedBlogposts.map((blogpost) => ({
      ...blogpost,
      key: blogpost.title,
      title: this.translateService.getTranslationKnowingTheyAreLoaded(
        blogpost.title
      )
    }));
  }
}
