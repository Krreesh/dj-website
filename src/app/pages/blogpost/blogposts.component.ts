import {ChangeDetectionStrategy, Component, OnInit, QueryList, ViewChildren} from '@angular/core';
import {Blogpost} from './blogpost';
import {BlogpostService} from './blogpost.service';
import {BehaviorSubject, combineLatest, Observable} from 'rxjs';
import {SortEvent} from '../data/sort.event';
import {SortableHeaderDirective} from '../data/sortable-header.directive';
import {map, tap} from 'rxjs/operators';
import {compare} from '../data/compare';

@Component({
  selector: 'my-blogposts',
  template: `
    <div class="container pt-4 pb-4">
      <h1>{{'website.blogposts' | translate}}</h1>
      <ng-container *ngIf="blogposts$ | async as blogposts">
        <table class="table table-striped table-responsive">
          <thead class="thead-dark">
          <tr>
            <th scope="col">#</th>
            <th scope="col" sortable="publicationDate" (sort)="onSort($event)">{{'data.blogpost.publicationDate' | translate}}</th>
            <th scope="col" sortable="title" (sort)="onSort($event)">{{'data.blogpost.title' | translate}}</th>
            <th scope="col" sortable="category" (sort)="onSort($event)">{{'data.blogpost.category' | translate}}</th>
          </tr>
          </thead>
          <tbody>
          <tr *ngFor="let blogpost of blogposts; index as zeroBasedRowNumber" blogpost-row
              [blogpost]="blogpost"
              [rowNumber]="zeroBasedRowNumber+1"></tr>
          </tbody>
        </table>
      </ng-container>
    </div>
  `,
  styles: [
      `
      .asc::before {
        content: "\\25be";
        float: right;
        color: gray;
      }

      .desc::before {
        content: "\\25b4";
        float: right;
        color: gray;
      }
    `
  ],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class BlogpostsComponent implements OnInit {
  blogposts$: Observable<Blogpost[]>;

  @ViewChildren(SortableHeaderDirective)
  sortableHeaderDirectives: QueryList<SortableHeaderDirective>;

  private sortEventBehaviorSubject = new BehaviorSubject<SortEvent>(SortEvent.unsortedEvent());

  constructor(private blogpostService: BlogpostService) {

  }

  ngOnInit(): void {
    const blogposts$ = this.blogpostService.getBlogposts$();

    const sortEventObservable$ = this.sortEventBehaviorSubject.asObservable().pipe(
      tap(sortEvent => this.resetHeadersToUnsorted(sortEvent))
    );

    this.blogposts$ = combineLatest([blogposts$, sortEventObservable$])
      .pipe(
        map(([blogposts, sortEvent]) => this.sortBlogposts(blogposts, sortEvent))
      );
  }

  private resetHeadersToUnsorted(sortEvent: SortEvent): void {
    if (!!this.sortableHeaderDirectives) {
      this.sortableHeaderDirectives
        .filter(header => !sortEvent.isEqualTo(header.sortable))
        .forEach(header => header.clearDirection());
    }
  }

  onSort(sortEvent: SortEvent): void {
    this.sortEventBehaviorSubject.next(sortEvent);
  }

  private sortBlogposts(blogposts: Blogpost[], sortEvent: SortEvent) {
    if (sortEvent.isUnsorted()) {
      return [...blogposts];
    }
    return this.sortBlogpostsAccordingToDirectionOfSortEvent(blogposts, sortEvent);
  }

  private sortBlogpostsAccordingToDirectionOfSortEvent(blogposts: Blogpost[], sortEvent: SortEvent): Blogpost[] {
    return [...blogposts].sort((blogpost1, blogpost2) => {
      const sortablePropertyName = sortEvent.sortablePropertyName;

      const blogpostProperty1 = blogpost1[sortablePropertyName];
      const blogpostProperty2 = blogpost2[sortablePropertyName];

      if (sortEvent.isAscending()) {
        return compare(blogpostProperty1, blogpostProperty2);
      }
      return compare(blogpostProperty2, blogpostProperty1);
    });
  }
}
