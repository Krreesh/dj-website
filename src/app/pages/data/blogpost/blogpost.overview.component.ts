import {Component, OnInit} from '@angular/core';
import {Blogpost} from './blogpost';
import {BlogpostService} from './blogpost.service';
import {Observable} from 'rxjs';

@Component({
  selector: 'blogpost-overview',
  template: `
    <div class="card" *ngIf="blogposts$ | async as blogposts">
      <div class="card-header">
        <h5 class="mb-0">
          <button type="button" class="btn btn-link btn-black" data-toggle="collapse" data-target="#blogpostContent"
                  aria-controls="blogpostContent" aria-label="Toggle blogpost content" [attr.aria-expanded]="!isCollapsed"
                  (click)="toggle()">
            <i class="fa" [ngClass]="isCollapsed ? 'fa-caret-down' : 'fa-caret-up'"></i>
            <span class="ml-2">{{'data.blogposts.description' | translate}}</span>
          </button>
        </h5>
      </div>
      <div class="card-body" id="blogpostContent" [ngbCollapse]="isCollapsed">
        <table class="table table-striped table-responsive">
          <thead class="thead-dark">
          <tr>
            <th scope="col">#</th>
            <th scope="col">{{'data.blogpost.title' | translate}}</th>
            <th scope="col">{{'data.blogpost.publicationDate' | translate}}</th>
            <th scope="col">{{'data.blogpost.category' | translate}}</th>
          </tr>
          </thead>
          <tbody>
          <tr *ngFor="let blogpost of blogposts ; index as i">
            <th scope="row">{{ i + 1 }}</th>
            <td>
              <ngb-highlight class="blogpost-title" [result]="blogpost.title"></ngb-highlight><a class="ml-1" target="_blank" href="{{blogpost.url}}"><i class="fa fa-external-link"></i></a>
            </td>
            <td>
              <ngb-highlight class="blogpost-publication-date" [result]="blogpost.publicationDate | dateLocale : 'DD MMMM YYYY'"></ngb-highlight>
            </td>
            <td>
              <ngb-highlight class="blogpost-category" [result]="blogpost.category"></ngb-highlight>
            </td>
          </tr>
          </tbody>
        </table>
      </div>
    </div>
  `,
  styles: [
      `
          .btn-black {
              color: black;
          }

          .btn-black:focus {
              text-decoration: none;
          }

          .btn-black:hover {
              text-decoration: underline;
          }
    `
  ]
})
export class BlogpostOverviewComponent implements OnInit {
  isCollapsed = true;

  blogposts$: Observable<Blogpost[]>;

  constructor(private blogpostService: BlogpostService) {

  }

  ngOnInit(): void {
    this.blogposts$ = this.blogpostService.getBlogposts$();
  }

  toggle(): void {
    this.isCollapsed = !this.isCollapsed;
  }
}
