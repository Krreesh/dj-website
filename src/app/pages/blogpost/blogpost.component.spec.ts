import {ComponentFixture, TestBed} from '@angular/core/testing';
import {TranslatePipeMock} from '../../translation/translate.pipe.mock';
import {BlogpostComponent} from './blogpost.component';
import {BlogpostOverviewComponent} from './blogpost-overview/blogpost.overview.component';
import {DateLocaleFilter} from '../../common/date.locale.filter';
import {NgbModule} from '@ng-bootstrap/ng-bootstrap';
import {HttpClientModule} from '@angular/common/http';
import {BlogpostService} from './blogpost-overview/blogpost.service';
import {TranslateModule} from '@ngx-translate/core';
import {MockComponent} from 'ng-mocks';

describe('BlogpostComponent', () => {
  let component: BlogpostComponent;
  let fixture: ComponentFixture<BlogpostComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [NgbModule, HttpClientModule, TranslateModule.forRoot()],
      declarations: [
        BlogpostComponent,
        MockComponent(BlogpostOverviewComponent),
        TranslatePipeMock,
        DateLocaleFilter,
      ],
      providers: [
        {
          provide: BlogpostService,
          useValue: {
            getBlogposts$: jest.fn(),
          },
        },
      ],
    });
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(BlogpostComponent);
    component = fixture.debugElement.componentInstance;
  });

  it('should render', () => {
    fixture.detectChanges();
    expect(fixture).toMatchInlineSnapshot(`
      <app-blogpost>
        <div
          class="container pt-4 pb-4"
        >
          <h1>
            blogpost.title.translated
          </h1>
          <blogpost-overview />
        </div>
      </app-blogpost>
    `);
  });
});
