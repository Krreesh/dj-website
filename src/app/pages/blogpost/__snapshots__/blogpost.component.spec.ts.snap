// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogpostComponent should render 1`] = `
<app-blogpost>
  <div
    class="container-fluid pt-4 pb-4"
  >
    <h1>
      blogpost.title.translated
    </h1>
    <blogpost-overview>
      
    </blogpost-overview>
  </div>
</app-blogpost>
`;
